ovn-org/ovn-kubernetes
# Add support for EgressIP and ExternalTrafficPolicy=Local to coexist

**- What this PR does and why is it needed**

This PR fixes the broken interaction between egressIPs and ETP=local features.
Currently if an EIP served pod is on a different node than the egressNode and
it is also a backend to the ETP=local service type, then the reply packet destined
for the ingress service traffic served by the EIP pod get's re-routed to the egressNode
using the reRoute policies we add for EIPs. This breaks the service traffic.

This PR implements the solution we did via https://issues.redhat.com/browse/FDP-42
to ensure these features work together seamlessly.

Solution:

Step1: Add a qos-rule to the node switches:

`from-lport   103 (ip4.src == $a4548040316634674295 && ct.trk && ct.rpl) mark=42`
where `a4548040316634674295` is the address set that contains all EIP pods.

Step2: Add new LRP at 102 to match on mark and simply "allow" thus skipping the re-routes
LRP:

sample:
```
102 pkt.mark == 42 allow
```

**However we must do the mark changes ONLY for local zone pods not for remote zone pods, else it breaks traffic since mark doesn't persist outside the local node.**

**- Special notes for reviewers**

- The logic to add QoS Rules has been added to the syncNamespaces/initClusterPolicies function so that on start all nodes get this rule
- It has also being added to the addNodeEvent handler so that when new node is added we get this rule
- It has not been added to updateNode event since there is nothing to do during node updates
- It has also not been added to the deleteNode event handler since switch goes away and rules will also go away unless it is still being referenced by another switch
- EgressSVC feature has not been added to this mix because there the reply traffic should be going out with SVC VIP so unsure what the expected behaviour for that feature is; but in the future if that is something that needs to be fixed it should be easy to do that the same way
- since its new LRP/QoS add it will take effect on upgrades

**- How to verify it**

1. All changes to existing unit tests already cover the unit testing around this
2. E2E has been added for cross functionality testing between services and egress ips.

The `2024-04-25T14:09:40.7078121Z [0mLoad Balancer Service Tests with MetalLB [0m[1mShould ensure load balancer service works when ETP=local and backend pods are also egressIP served pods[0m` is running on both v4 and v6 lanes for LGW gateway mode and here the egressNode is different from the node where the pod lives

AND `should work on secondary node interfaces for ETP=local and ETP=cluster when backend pods are also served by EgressIP` is running on SGW mode both v4 & v6

**- Description for the changelog**
`Support ETP=local with EIP`

/label ci
